src/pip/_internal/download.py

```diff
@@ -138,31 +143,79 @@ def __init__(self, prompting=True):
+    def _get_url_and_credentials(self, original_url,
+                                 allow_netrc=True, allow_keyring=True):
+        """Returns the credentials to use for the provided URL.
+
+        If allowed, netrc and keyring may be used to obtain the
+        correct credentials.
+
+        Returns (url_without_credentials, username, password). Note
+        that even if the original URL contains credentials, the result
+        of this function may include different values.
+        """
+        parsed = urllib_parse.urlparse(original_url)
 
         # Split the credentials from the netloc.
         netloc, url_user_password = split_auth_from_netloc(parsed.netloc)
 
-        # Set the url of the request to the url without any credentials
-        req.url = urllib_parse.urlunparse(parsed[:1] + (netloc,) + parsed[2:])
+        # Return the url without any credentials
+        url = urllib_parse.urlunparse(parsed[:1] + (netloc,) + parsed[2:])
 
         # Use any stored credentials that we have for this netloc
         username, password = self.passwords.get(netloc, (None, None))
 
         # Use the credentials embedded in the url if we have none stored
         if username is None:
             username, password = url_user_password
+            if username or password:
+                logger.debug("Found credentials in url for %s", netloc)
 
         # Get creds from netrc if we still don't have them
-        if username is None and password is None:
-            netrc_auth = get_netrc_auth(req.url)
+        if allow_netrc and username is None and password is None:
+            netrc_auth = get_netrc_auth(original_url)
+            if netrc_auth:
+                logger.debug("Found credentials in netrc for %s", netloc)
             username, password = netrc_auth if netrc_auth else (None, None)
 
+        # If available, use keyring. If we have a username, use it, but
+        # allow keyring to provide one based on the URL.
+        if allow_keyring and password is None and keyring is not None:
+            try:
+                try:
+                    get_credential = keyring.get_credential
+                except AttributeError:
+                    logger.debug("Getting password from keyring for %s",
+                                 netloc)
+                    password = keyring.get_password(url, username)
+                    if password:
+                        logger.debug("Found password in keyring")
+                else:
+                    logger.debug("Getting credentials from keyring for %s",
+                                 netloc)
+                    cred = get_credential(url, username)
+                    if cred is not None:
+                        logger.debug("Found credentials in keyring")
+                        username = cred.username
+                        password = cred.password
+            except Exception as exc:
+                logger.warn("Keyring is skipped due to an exception: %s",
```

Comment:
`warn` is deprecated in favor of `logger.warning`
```suggestion
                logger.warning("Keyring is skipped due to an exception: %s",
```